### PR TITLE
WFLY-14953 distributable-ejb subsystem

### PR DESCRIPTION
**Pull request by pferraro**
https://issues.redhat.com/browse/WFLY-14953

Analysis document: https://github.com/wildfly/wildfly-proposals/pull/427

Submitted on behalf of @rachmatowicz 

Commits expected to be squashed before merge.

### CODE REVIEW

File: clustering/ejb/extension/pom.xml [hunk 1]
```diff
@@ -0,0 +1,71 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!--
+  ~ JBoss, Home of Professional Open Source.
+  ~ Copyright 2018, Red Hat, Inc., and individual contributors
```
Comment (pferraro) [inline clustering/ejb/extension/pom.xml:4]: Update year

File: clustering/ejb/extension/src/main/java/org/wildfly/extension/clustering/ejb/ClientMappingsRegistryProviderResourceDefinition.java [hunk 2]
```diff
@@ -0,0 +1,70 @@
+package org.wildfly.extension.clustering.ejb;
```
Comment (pferraro) [inline clustering/ejb/extension/src/main/java/org/wildfly/extension/clustering/ejb/ClientMappingsRegistryProviderResourceDefinition.java:1]: Missing copyright header.

File: clustering/ejb/extension/src/main/java/org/wildfly/extension/clustering/ejb/ClientMappingsRegistryProviderServiceConfigurator.java [hunk 3]
```diff
@@ -0,0 +1,38 @@
+package org.wildfly.extension.clustering.ejb;
```
Comment (pferraro) [inline clustering/ejb/extension/src/main/java/org/wildfly/extension/clustering/ejb/ClientMappingsRegistryProviderServiceConfigurator.java:1]: Missing copyright header.

File: clustering/ejb/extension/src/main/java/org/wildfly/extension/clustering/ejb/InfinispanClientMappingsRegistryProviderResourceDefinition.java [hunk 4]
```diff
@@ -0,0 +1,61 @@
+package org.wildfly.extension.clustering.ejb;
```
Comment (pferraro) [inline clustering/ejb/extension/src/main/java/org/wildfly/extension/clustering/ejb/InfinispanClientMappingsRegistryProviderResourceDefinition.java:1]: Missing copyright header.

File: clustering/ejb/extension/src/main/java/org/wildfly/extension/clustering/ejb/InfinispanClientMappingsRegistryProviderServiceConfigurator.java [hunk 5]
```diff
@@ -0,0 +1,40 @@
+package org.wildfly.extension.clustering.ejb;
```
Comment (pferraro) [inline clustering/ejb/extension/src/main/java/org/wildfly/extension/clustering/ejb/InfinispanClientMappingsRegistryProviderServiceConfigurator.java:1]: Missing copyright header.

File: clustering/ejb/extension/src/main/java/org/wildfly/extension/clustering/ejb/LocalClientMappingsRegistryProviderResourceDefinition.java [hunk 6]
```diff
@@ -0,0 +1,20 @@
+package org.wildfly.extension.clustering.ejb;
```
Comment (pferraro) [inline clustering/ejb/extension/src/main/java/org/wildfly/extension/clustering/ejb/LocalClientMappingsRegistryProviderResourceDefinition.java:1]: Missing copyright header.

File: clustering/ejb/extension/src/main/java/org/wildfly/extension/clustering/ejb/LocalClientMappingsRegistryProviderServiceConfigurator.java [hunk 7]
```diff
@@ -0,0 +1,22 @@
+package org.wildfly.extension.clustering.ejb;
```
Comment (pferraro) [inline clustering/ejb/extension/src/main/java/org/wildfly/extension/clustering/ejb/LocalClientMappingsRegistryProviderServiceConfigurator.java:1]: Missing copyright header.

File: clustering/ejb/extension/src/main/resources/schema/wildfly-distributable-ejb_1_0.xsd [hunk 8]
```diff
@@ -0,0 +1,99 @@
+<?xml version="1.0" encoding="UTF-8"?>
+
+<!--
+  ~ JBoss, Home of Professional Open Source.
+  ~ Copyright 2018, Red Hat, Inc., and individual contributors
```
Comment (pferraro) [inline clustering/ejb/extension/src/main/resources/schema/wildfly-distributable-ejb_1_0.xsd:5]: Fix year.

File: clustering/ejb/extension/src/main/java/org/wildfly/extension/clustering/ejb/InfinispanBeanManagementServiceConfigurator.java [hunk 9]
```diff
@@ -0,0 +1,105 @@
+/*
+ * JBoss, Home of Professional Open Source.
+ * Copyright 2018, Red Hat, Inc., and individual contributors
```
Comment (pferraro) [inline clustering/ejb/extension/src/main/java/org/wildfly/extension/clustering/ejb/InfinispanBeanManagementServiceConfigurator.java:3]: Fix year.

File: clustering/ejb/extension/src/main/java/org/wildfly/extension/clustering/ejb/LocalClientMappingsRegistryProvider.java [hunk 10]
```diff
@@ -0,0 +1,72 @@
+/*
+ * JBoss, Home of Professional Open Source.
+ * Copyright 2020, Red Hat, Inc., and individual contributors
```
Comment (pferraro) [inline clustering/ejb/extension/src/main/java/org/wildfly/extension/clustering/ejb/LocalClientMappingsRegistryProvider.java:3]: Fix year.

File: clustering/ejb/extension/src/test/java/org/wildfly/extension/clustering/ejb/DistributableEjbSubsystemTestCase.java [hunk 11]
```diff
@@ -0,0 +1,64 @@
+/*
+ * JBoss, Home of Professional Open Source.
+ * Copyright 2018, Red Hat, Inc., and individual contributors
```
Comment (pferraro) [inline clustering/ejb/extension/src/test/java/org/wildfly/extension/clustering/ejb/DistributableEjbSubsystemTestCase.java:3]: Fix year

File: clustering/ejb/spi/src/main/java/org/wildfly/clustering/ejb/EjbRequirement.java [hunk 12]
```diff
@@ -0,0 +1,35 @@
+package org.wildfly.clustering.ejb;
```
Comment (pferraro) [inline clustering/ejb/spi/src/main/java/org/wildfly/clustering/ejb/EjbRequirement.java:1]: Missing copyright header.

File: ejb3/src/main/java/org/jboss/as/ejb3/cache/distributable/LegacyDistributableCacheFactoryBuilderServiceConfigurator.java [hunk 13]
```diff
@@ -0,0 +1,66 @@
+/*
+ * JBoss, Home of Professional Open Source.
+ * Copyright 2013, Red Hat, Inc., and individual contributors
```
Comment (pferraro) [inline ejb3/src/main/java/org/jboss/as/ejb3/cache/distributable/LegacyDistributableCacheFactoryBuilderServiceConfigurator.java:3]: Fix year.

File: ejb3/src/main/java/org/jboss/as/ejb3/subsystem/LegacyCacheFactoryResourceDefinition.java [hunk 14]
```diff
@@ -0,0 +1,83 @@
+/*
+ * JBoss, Home of Professional Open Source.
+ * Copyright 2011, Red Hat, Inc., and individual contributors
```
Comment (pferraro) [inline ejb3/src/main/java/org/jboss/as/ejb3/subsystem/LegacyCacheFactoryResourceDefinition.java:3]: Fix year

File: ejb3/src/main/java/org/jboss/as/ejb3/subsystem/EJBTransformers.java [hunk 15]
```diff
@@ -0,0 +1,70 @@
...
+import org.jboss.as.controller.transform.description.ResourceTransformationDescriptionBuilder;
+import org.jboss.as.controller.transform.description.TransformationDescriptionBuilder;
+
+/**
+ * Jakarta Enterprise Beans Transformers used to transform current model version to legacy model versions for domain mode.
+ *
+ * @author Tomaz Cerar (c) 2017 Red Hat Inc.
+ * @author Richard Achmatowicz (c) 2020 Red Hat Inc.
+ */
+public class EJBTransformers implements ExtensionTransformerRegistration {
```
Comment (pferraro) [inline ejb3/src/main/java/org/jboss/as/ejb3/subsystem/EJBTransformers.java:40]: Needs `@MetaInfServices(ExtensionTransformerRegistration.class)` annotation

**Review by pferraro — COMMENTED** [review]
@rachmatowicz See requested changes.

File: ee-feature-pack/common/src/main/resources/modules/system/layers/base/org/wildfly/extension/clustering/ejb/main/module.xml [hunk 16]
```diff
@@ -0,0 +1,59 @@
...
+        <module name="org.jboss.msc"/>
+        <module name="org.jboss.staxmapper"/>
+        <module name="org.jboss.vfs"/>
+        <module name="org.wildfly.clustering.ee.spi"/>
+        <module name="org.wildfly.clustering.ejb.infinispan"/>
+        <module name="org.wildfly.clustering.ejb.spi"/>
+        <module name="org.wildfly.clustering.infinispan.client"/>
+        <module name="org.wildfly.clustering.infinispan.spi"/>
+        <module name="org.wildfly.clustering.service"/>
+        <module name="org.wildfly.clustering.spi"/>
```
Comment (yersan) [inline ee-feature-pack/common/src/main/resources/modules/system/layers/base/org/wildfly/extension/clustering/ejb/main/module.xml:57]: Could you justify the uses of the following modules as dependencies? A inspection of `org.wildfly:wildfly-clustering-ejb-extension` reveals they are not used, so I believe you could be adding more dependencies than needed:

* java.xml
* org.infinispan
* org.jboss.as.ee
* org.jboss.as.server
* org.jboss.as.web-common
* org.jboss.logging
* org.jboss.metadata.common
* org.jboss.metadata.web
* org.jboss.vfs
* org.wildfly.clustering.infinispan.client

That could affect the trimming capabilities by provisioning more dependencies than needed, so it is worth double-checking them; there could be details that are not revealed by inspecting only the artifact itself, so I'm just asking.

Comment (rachmatowicz) [reply ee-feature-pack/common/src/main/resources/modules/system/layers/base/org/wildfly/extension/clustering/ejb/main/module.xml]: Thanks, Yeray. I incorporated your changes.

File: clustering/ejb/extension/pom.xml (hunk 1 above)
Comment (rachmatowicz) [reply clustering/ejb/extension/pom.xml]: Done.

File: clustering/ejb/extension/src/main/java/org/wildfly/extension/clustering/ejb/ClientMappingsRegistryProviderResourceDefinition.java (hunk 2 above)
Comment (rachmatowicz) [reply clustering/ejb/extension/src/main/java/org/wildfly/extension/clustering/ejb/ClientMappingsRegistryProviderResourceDefinition.java:22]: Done.

File: clustering/ejb/extension/src/main/java/org/wildfly/extension/clustering/ejb/ClientMappingsRegistryProviderServiceConfigurator.java (hunk 3 above)
Comment (rachmatowicz) [reply clustering/ejb/extension/src/main/java/org/wildfly/extension/clustering/ejb/ClientMappingsRegistryProviderServiceConfigurator.java:22]: Done.

File: clustering/ejb/extension/src/main/java/org/wildfly/extension/clustering/ejb/InfinispanClientMappingsRegistryProviderResourceDefinition.java (hunk 4 above)
Comment (rachmatowicz) [reply clustering/ejb/extension/src/main/java/org/wildfly/extension/clustering/ejb/InfinispanClientMappingsRegistryProviderResourceDefinition.java:22]: Done.

File: clustering/ejb/extension/src/main/java/org/wildfly/extension/clustering/ejb/InfinispanClientMappingsRegistryProviderServiceConfigurator.java (hunk 5 above)
Comment (rachmatowicz) [reply clustering/ejb/extension/src/main/java/org/wildfly/extension/clustering/ejb/InfinispanClientMappingsRegistryProviderServiceConfigurator.java:22]: Done.

File: clustering/ejb/extension/src/main/java/org/wildfly/extension/clustering/ejb/LocalClientMappingsRegistryProviderResourceDefinition.java (hunk 6 above)
Comment (rachmatowicz) [reply clustering/ejb/extension/src/main/java/org/wildfly/extension/clustering/ejb/LocalClientMappingsRegistryProviderResourceDefinition.java:22]: Done.

File: clustering/ejb/extension/src/main/java/org/wildfly/extension/clustering/ejb/LocalClientMappingsRegistryProviderServiceConfigurator.java (hunk 7 above)
Comment (rachmatowicz) [reply clustering/ejb/extension/src/main/java/org/wildfly/extension/clustering/ejb/LocalClientMappingsRegistryProviderServiceConfigurator.java:22]: Done.

File: clustering/ejb/extension/src/main/resources/schema/wildfly-distributable-ejb_1_0.xsd (hunk 8 above)
Comment (rachmatowicz) [reply clustering/ejb/extension/src/main/resources/schema/wildfly-distributable-ejb_1_0.xsd]: Done.

File: clustering/ejb/extension/src/main/java/org/wildfly/extension/clustering/ejb/InfinispanBeanManagementServiceConfigurator.java (hunk 9 above)
Comment (rachmatowicz) [reply clustering/ejb/extension/src/main/java/org/wildfly/extension/clustering/ejb/InfinispanBeanManagementServiceConfigurator.java]: Done.

File: clustering/ejb/extension/src/main/java/org/wildfly/extension/clustering/ejb/LocalClientMappingsRegistryProvider.java (hunk 10 above)
Comment (rachmatowicz) [reply clustering/ejb/extension/src/main/java/org/wildfly/extension/clustering/ejb/LocalClientMappingsRegistryProvider.java]: Done.

File: clustering/ejb/extension/src/test/java/org/wildfly/extension/clustering/ejb/DistributableEjbSubsystemTestCase.java (hunk 11 above)
Comment (rachmatowicz) [reply clustering/ejb/extension/src/test/java/org/wildfly/extension/clustering/ejb/DistributableEjbSubsystemTestCase.java]: Done.

File: clustering/ejb/spi/src/main/java/org/wildfly/clustering/ejb/EjbRequirement.java (hunk 12 above)
Comment (rachmatowicz) [reply clustering/ejb/spi/src/main/java/org/wildfly/clustering/ejb/EjbRequirement.java:22]: Done.

File: ejb3/src/main/java/org/jboss/as/ejb3/cache/distributable/LegacyDistributableCacheFactoryBuilderServiceConfigurator.java (hunk 13 above)
Comment (rachmatowicz) [reply ejb3/src/main/java/org/jboss/as/ejb3/cache/distributable/LegacyDistributableCacheFactoryBuilderServiceConfigurator.java]: Done.

File: ejb3/src/main/java/org/jboss/as/ejb3/subsystem/LegacyCacheFactoryResourceDefinition.java (hunk 14 above)
Comment (rachmatowicz) [reply ejb3/src/main/java/org/jboss/as/ejb3/subsystem/LegacyCacheFactoryResourceDefinition.java]: Done.

File: ejb3/src/main/java/org/jboss/as/ejb3/subsystem/EJBTransformers.java (hunk 15 above)
Comment (rachmatowicz) [reply ejb3/src/main/java/org/jboss/as/ejb3/subsystem/EJBTransformers.java:42]: Done.

**Comment by rachmatowicz** [comment]
retest please

File: ee-feature-pack/common/src/main/resources/modules/system/layers/base/org/wildfly/extension/clustering/ejb/main/module.xml (hunk 16 above)
Comment (yersan) [reply ee-feature-pack/common/src/main/resources/modules/system/layers/base/org/wildfly/extension/clustering/ejb/main/module.xml]: You're welcome @rachmatowicz , the module looks good to me now.

**Review by yersan — COMMENTED** [review]
@rachmatowicz About the HostExclude test case, you are adding a new extension, so you need to add such an extension to the HostExclude test case so we can track and verify the host exclusions. 

Add it [here](https://github.com/wildfly/wildfly/blob/edacac3c85baec7f935655bd4f5164a839328726/testsuite/domain/src/test/java/org/jboss/as/test/integration/domain/HostExcludesTestCase.java#L201). something like this:

```
CURRENT(MAJOR, WILDFLY_26_0, Arrays.asList(
                "org.wildfly.extension.clustering.ejb"
        ));
```

**Comment by rachmatowicz** [comment]
Couple of failures with SingletonDeploymentJBossAllTestCase.

**Comment by pferraro** [comment]
Closing this so it can be reopened by @rachmatowicz

**Comment by rachmatowicz** [comment]
The PR has been moved from here to: https://github.com/wildfly/wildfly/pull/15162

**Pull request closed without merging**